core/src/main/java/hudson/TcpSlaveAgentListener.java

```diff
@@ -200,6 +210,14 @@ public ConnectionHandler(Socket s) {
                 id = iotaGen++;
             }
             setName("TCP agent connection handler #"+id+" with "+s.getRemoteSocketAddress());
+            setUncaughtExceptionHandler((t, e) -> {
+                LOGGER.log(Level.SEVERE, "Uncaught exception in TcpSlaveAgentListener ConnectionHandler " + t, e);
+                try {
+                    s.close();
```

Comment:
It should also abort and restart the TcpSlaveAgentListener thread, no? Or at least to make the listener to restart the handler via a callback